File: packages/ruby/lib/identifold/storage/postgres.rb
require "pg"
require_relative "../../identifold"
require_relative "../storage"

module Identifold
  module Storage
    class PostgresAdapter
      def initialize(connection)
        @connection = connection
      end

      def reserve(request)
        result = query(
          "SELECT identifold_reserve_reference($1::text::uuid, $2::text, $3::text) AS reserved",
          [request.machine_id, request.namespace, request.reference]
        )
        value = result.first&.fetch("reserved", nil)
        return true if value == "t"
        return false if value == "f"

        raise Error, "allocation_conflict"
      end

      def resolve(reference, namespace)
        result = query(
          "SELECT resolved_machine_id::text AS machine_id, resolved_namespace AS namespace " \
            "FROM identifold_resolve_reference($1::text, $2::text)",
          [reference, namespace]
        )
        return nil if result.ntuples.zero?
        raise Error, "allocation_conflict" unless result.ntuples == 1

        row = result.first
        ReferenceMapping.new(machine_id: row.fetch("machine_id"), namespace: row.fetch("namespace"))
      end

      def allocate(request)
        result = query(
          "SELECT identifold_allocate_sequence(" \
            "$1::text::uuid, $2::text, $3::text, $4::text, $5::smallint) AS sequence",
          [
            request.machine_id,
            request.namespace,
            request.reference_prefix,
            request.scope,
            request.width
          ]
        )
        value = result.first&.fetch("sequence", nil)
        raise Error, "allocation_conflict" unless value&.match?(/\A\d+\z/)

        value.to_i
      end

      private

      def query(sql, values)
        @connection.exec_params(sql, values)
      rescue PG::Error => exception
        code = exception.result&.error_field(PG::Result::PG_DIAG_SQLSTATE)
        raise Error, case code
                     when "22003" then "sequence_overflow"
                     when "22023" then "invalid_allocation_policy"
                     else "allocation_conflict"
                     end
      end
    end
  end
end

File: packages/python/src/identifold/storage/postgres.py
from __future__ import annotations

from typing import Any, Protocol

from identifold import IdentifoldError

from .base import (
    ReferenceMapping,
    ReferenceReservation,
    SequenceAllocationRequest,
    StorageAdapter,
)


class AsyncCursor(Protocol):
    async def fetchone(self) -> tuple[Any, ...] | None: ...


class AsyncPostgresConnection(Protocol):
    async def execute(
        self, query: str, params: tuple[object, ...] = ()
    ) -> AsyncCursor: ...


class PostgresStorageAdapter(StorageAdapter):
    """Storage adapter for a caller-owned psycopg-compatible async connection."""

    def __init__(self, connection: AsyncPostgresConnection) -> None:
        self._connection = connection

    async def reserve(self, request: ReferenceReservation) -> bool:
        row = await self._query_one(
            "SELECT identifold_reserve_reference(%s::uuid, %s::text, %s::text)",
            (request.machine_id, request.namespace, request.reference),
        )
        if row is None or len(row) != 1 or not isinstance(row[0], bool):
            raise IdentifoldError(
                "allocation_conflict",
                "Reference reservation returned an invalid result",
            )
        return row[0]

    async def resolve(self, reference: str, namespace: str) -> ReferenceMapping | None:
        row = await self._query_one(
            "SELECT resolved_machine_id::text, resolved_namespace "
            "FROM identifold_resolve_reference(%s::text, %s::text)",
            (reference, namespace),
        )
        if row is None:
            return None
        if len(row) != 2 or not isinstance(row[0], str) or not isinstance(row[1], str):
            raise IdentifoldError(
                "allocation_conflict", "Reference lookup returned an invalid result"
            )
        return ReferenceMapping(row[0], row[1])

    async def allocate(self, request: SequenceAllocationRequest) -> int:
        row = await self._query_one(
            "SELECT identifold_allocate_sequence(%s::uuid, %s::text, %s::text, %s::text, %s::smallint)",
            (
                request.machine_id,
                request.namespace,
                request.reference_prefix,
                request.scope,
                request.width,
            ),
        )
        if (
            row is None
            or len(row) != 1
            or isinstance(row[0], bool)
            or not isinstance(row[0], int)
            or row[0] < 0
        ):
            raise IdentifoldError(
                "allocation_conflict", "Sequence allocation returned an invalid result"
            )
        return row[0]

    async def _query_one(
        self, query: str, params: tuple[object, ...]
    ) -> tuple[Any, ...] | None:
        try:
            cursor = await self._connection.execute(query, params)
            return await cursor.fetchone()
        except IdentifoldError:
            raise
        except Exception as error:  # noqa: BLE001 - drivers expose distinct exception trees
            sqlstate = getattr(error, "sqlstate", None)
            if sqlstate == "22003":
                raise IdentifoldError(
                    "sequence_overflow", "Sequential reference capacity is exhausted"
                ) from None
            if sqlstate == "22023":
                raise IdentifoldError(
                    "invalid_allocation_policy",
                    "Sequential allocation policy is invalid",
                ) from None
            raise IdentifoldError(
                "allocation_conflict", "Database operation could not be committed"
            ) from None
